File: packages/xyd-opensdk-framework/src/example-plan.ts
// Shared example-value planner for the generated SDK test suites (generateTests).
// It turns a TypeRef into a language-NEUTRAL example tree; each emitter renders
// that tree in its own syntax (Go composite literals, Python kwargs). This keeps
// "what a realistic example value is" in one place — the same way planOperation
// centralizes request semantics — so the Go and Python test suites exercise the
// SAME shapes and can never drift.
import type { Method, NamedType, Param, TypeRef } from '@xyd-js/opensdk-core';

/** A language-neutral example value; emitters render it to their own syntax. */
export type ExampleValue =
  | { kind: 'string'; value: string }
  | { kind: 'integer'; value: number }
  | { kind: 'number'; value: number }
  | { kind: 'boolean'; value: boolean }
  | { kind: 'null' }
  | { kind: 'binary' }
  | { kind: 'enum'; typeName: string; value: unknown }
  | { kind: 'const'; value: unknown }
  | { kind: 'array'; item: ExampleValue }
  | { kind: 'map'; value: ExampleValue }
  | { kind: 'object'; typeName?: string; fields: ExampleField[] }
  | { kind: 'union'; typeName: string; variant: ExampleValue }
  | { kind: 'any' };

/** One example field of an object/params example. */
export interface ExampleField {
  name: string;
  value: ExampleValue;
  required: boolean;
}

export interface PlanExampleOptions {
  /** Include optional struct fields (the "with all params" / "WithOptionalParams" example). */
  withOptional?: boolean;
  /** Example text for a bare string scalar at THIS level (not propagated into nested types). */
  stringHint?: string;
  /**
   * Prefer the spec's own `example`/`default` values and emit format-aware scalar
   * samples (date-time, uuid, email, ...) instead of the neutral `0`/`'x'`. OFF by
   * default so the generated SDK TEST goldens + MockServer responses stay stable;
   * only the doc USAGE snippets (`generateUsage`) opt in.
   */
  realistic?: boolean;
}

/** The example call for one method — everything a test needs to invoke it. */
export interface MethodExample {
  /** Positional path arguments, in path order (each with its source param). */
  pathArgs: { param: Param; value: ExampleValue }[];
  /** Params-struct / keyword fields: query ∪ header ∪ request-body fields. */
  fields: ExampleField[];
  /** True when any non-required field exists — drives the "with all params" variant. */
  hasOptional: boolean;
}

const MAX_DEPTH = 6;

/**
 * Resolve a TypeRef to an example value, expanding named types via the symbol
 * table. Cycle-guarded (a type currently being expanded yields a terminal empty
 * object) and depth-capped so deeply/recursively nested schemas stay finite.
 */
export function planExample(
  ref: TypeRef | undefined,
  types: Map<string, NamedType>,
  opts: PlanExampleOptions = {},
  seen: Set<string> = new Set(),
  depth = 0,
): ExampleValue {
  if (!ref || depth > MAX_DEPTH) return { kind: 'any' };

  if (ref.const !== undefined) return { kind: 'const', value: ref.const };

  switch (ref.kind) {
    case 'scalar':
      return scalarExample(ref, opts.stringHint, opts.realistic);
    case 'array':
      return { kind: 'array', item: planExample(ref.items as TypeRef, types, dropHint(opts), seen, depth + 1) };
    case 'map':
      return { kind: 'map', value: planExample(ref.values as TypeRef, types, dropHint(opts), seen, depth + 1) };
    case 'any':
      return { kind: 'any' };
    case 'ref':
      return refExample(ref, types, opts, seen, depth);
    default:
      return { kind: 'any' };
  }
}

function scalarExample(ref: TypeRef, stringHint?: string, realistic?: boolean): ExampleValue {
  const fmt = (ref.format || '').toLowerCase();
  if (fmt === 'binary') return { kind: 'binary' };
  switch (ref.scalar) {
    case 'integer':
      return { kind: 'integer', value: realistic ? 1 : 0 };
    case 'number':
      return { kind: 'number', value: realistic ? 1 : 0 };
    case 'boolean':
      return { kind: 'boolean', value: true };
    default:
      return { kind: 'string', value: realistic ? realisticString(fmt, stringHint) : (stringHint ?? 'x') };
  }
}

/** A believable string sample for a known `format`, else the hint (a param name
 * reads well, e.g. `After: "after"`) or a neutral `"string"`. `realistic` only. */
function realisticString(fmt: string, hint?: string): string {
  switch (fmt) {
    case 'date-time':
      return '2024-01-01T00:00:00Z';
    case 'date':
      return '2024-01-01';
    case 'email':
      return 'user@example.com';
    case 'uri':
    case 'url':
      return 'https://example.com';
    case 'uuid':
      return '123e4567-e89b-12d3-a456-426614174000';
    case 'hostname':
      return 'example.com';
    case 'ipv4':
      return '192.0.2.1';
    default:
      return hint ?? 'string';
  }
}

/**
 * Coerce a spec `example`/`default` JSON literal into a neutral ExampleValue —
 * ONLY for scalar/scalar-array shapes (a struct/enum/union literal would need the
 * type to render correctly, so we return undefined and let planExample walk the
 * type). Returns undefined when it can't safely coerce.
 */
function literalToExample(value: unknown): ExampleValue | undefined {
  switch (typeof value) {
    case 'string':
      return { kind: 'string', value };
    case 'boolean':
      return { kind: 'boolean', value };
    case 'number':
      return Number.isInteger(value) ? { kind: 'integer', value } : { kind: 'number', value };
  }
  if (value === null) return { kind: 'null' };
  if (Array.isArray(value) && value.length) {
    const item = literalToExample(value[0]);
    if (item) return { kind: 'array', item };
  }
  return undefined;
}

/** Under `realistic`, the spec's own example/default for a scalar/array-typed
 * param or field (skips ref types — enum/struct/union render from the type).
 * Exported so re-planning emitters (Go, .NET, which walk fields via `planExample`
 * rather than `planMethodExample.fields`) can honor spec example/default too. */
export function realisticLiteral(type: TypeRef | undefined, ...candidates: unknown[]): ExampleValue | undefined {
  if (!type || (type.kind !== 'scalar' && type.kind !== 'array')) return undefined;
  for (const candidate of candidates) {
    if (candidate === undefined) continue;
    const example = literalToExample(candidate);
    if (example) return example;
  }
  return undefined;
}

function refExample(
  ref: TypeRef,
  types: Map<string, NamedType>,
  opts: PlanExampleOptions,
  seen: Set<string>,
  depth: number,
): ExampleValue {
  const name = ref.name;
  const named = name ? types.get(name) : undefined;
  if (!named || !name) return { kind: 'any' };

  // cycle guard: a self/mutually-referential struct yields a terminal example
  if (seen.has(name)) return { kind: 'object', typeName: name, fields: [] };

  switch (named.kind) {
    case 'enum': {
      const first = named.values?.[0];
      return { kind: 'enum', typeName: name, value: first ? first.value : '' };
    }
    case 'alias':
      return planExample(named.of as TypeRef, types, dropHint(opts), seen, depth + 1);
    case 'union': {
      const variant = (named.variants || [])[0] as TypeRef | undefined;
      const nested = new Set(seen).add(name);
      return { kind: 'union', typeName: name, variant: planExample(variant, types, dropHint(opts), nested, depth + 1) };
    }
    default: {
      // struct: fill required fields; optional too when withOptional
      const nested = new Set(seen).add(name);
      const fields = exampleFields(named.fields || [], types, opts, nested, depth + 1);
      return { kind: 'object', typeName: name, fields };
    }
  }
}

/** Example fields for a struct's field list; required-first, optionals only when withOptional. */
export function exampleFields(
  fields: { name: string; type: TypeRef; required?: boolean; default?: unknown }[],
  types: Map<string, NamedType>,
  opts: PlanExampleOptions,
  seen: Set<string> = new Set(),
  depth = 0,
): ExampleField[] {
  const out: ExampleField[] = [];
  const wanted = fields.filter((f) => f.required || opts.withOptional);
  // required first, then optional — deterministic, readable literals
  wanted.sort((a, b) => Number(!!b.required) - Number(!!a.required));
  const nested: PlanExampleOptions = { withOptional: opts.withOptional, realistic: opts.realistic };
  for (const f of wanted) {
    // Field carries only `default` (no `example`); prefer it under `realistic`.
    const value =
      (opts.realistic ? realisticLiteral(f.type, f.default) : undefined) ??
      planExample(f.type, types, nested, seen, depth);
    out.push({ name: f.name, required: !!f.required, value });
  }
  return out;
}

/** Drop the per-level stringHint before recursing into nested types (keep withOptional + realistic). */
function dropHint(opts: PlanExampleOptions): PlanExampleOptions {
  return opts.stringHint === undefined ? opts : { withOptional: opts.withOptional, realistic: opts.realistic };
}

/**
 * The example call for a method: positional path args + the params-struct /
 * keyword fields (query ∪ header ∪ request-body fields). Path/query/header
 * string examples use the param name (readable, non-empty like openai-go's
 * `After: "after"`); nested body strings default to "x".
 */
export function planMethodExample(
  method: Method,
  types: Map<string, NamedType>,
  opts: { withOptional?: boolean; realistic?: boolean } = {},
): MethodExample {
  const withOptional = !!opts.withOptional;
  const realistic = !!opts.realistic;

  const pathArgs = (method.pathParams || []).map((param) => ({
    param,
    value:
      (realistic ? realisticLiteral(param.type as TypeRef, param.example, param.default) : undefined) ??
      planExample(param.type as TypeRef, types, { stringHint: param.name, realistic }),
  }));

  const fields: ExampleField[] = [];
  const addParam = (p: Param) => {
    if (!p.required && !withOptional) return;
    fields.push({
      name: p.name,
      required: !!p.required,
      // A param carries both `example` and `default`; prefer them under `realistic`.
      value:
        (realistic ? realisticLiteral(p.type as TypeRef, p.example, p.default) : undefined) ??
        planExample(p.type as TypeRef, types, { withOptional, stringHint: p.name, realistic }),
    });
  };
  for (const p of method.queryParams || []) addParam(p);
  for (const p of method.headerParams || []) addParam(p);

  // request body: flatten its struct fields into the params struct / kwargs
  const bodyRef = method.requestBody?.type as TypeRef | undefined;
  if (bodyRef?.kind === 'ref' && bodyRef.name) {
    const named = types.get(bodyRef.name);
    if (named?.kind === 'struct') {
      for (const f of exampleFields(named.fields || [], types, { withOptional, realistic })) fields.push(f);
    }
  }

  // required-first for stable, readable output
  fields.sort((a, b) => Number(b.required) - Number(a.required));

  const hasOptional =
    (method.queryParams || []).some((p) => !p.required) ||
    (method.headerParams || []).some((p) => !p.required) ||
    bodyHasOptional(bodyRef, types);

  return { pathArgs, fields, hasOptional };
}

function bodyHasOptional(bodyRef: TypeRef | undefined, types: Map<string, NamedType>): boolean {
  if (bodyRef?.kind !== 'ref' || !bodyRef.name) return false;
  const named = types.get(bodyRef.name);
  return !!named?.fields?.some((f) => !f.required);
}

File: crates/xyd_opensdk_ruby/src/example_plan.rs
//! Port of the framework `example-plan.ts`: the shared, language-neutral example
//! -value planner the generated test suites use. It turns a TypeRef into an
//! `ExampleValue` tree; `example.rs` renders that tree into Ruby literals, so the
//! Go/Python/Ruby suites exercise identical shapes and can never drift.

use std::collections::{HashMap, HashSet};

use serde_json::Value;

/// A language-neutral example value; `example.rs` renders it to Ruby syntax.
pub enum ExampleValue {
    Str(String),
    Integer(i64),
    Number(f64),
    Boolean(bool),
    Null,
    Binary,
    Enum(Value),
    Const(Value),
    Array(Box<ExampleValue>),
    Map(Box<ExampleValue>),
    Object(Vec<ExampleField>),
    Union(Box<ExampleValue>),
    Any,
}

/// One example field of an object/params example.
pub struct ExampleField {
    pub name: String,
    pub required: bool,
    pub value: ExampleValue,
}

/// One positional path argument, with its param name (for the guard test).
pub struct PathArg {
    pub name: String,
    pub value: ExampleValue,
}

/// The example call for one method — everything a test needs to invoke it.
pub struct MethodExample {
    pub path_args: Vec<PathArg>,
    pub fields: Vec<ExampleField>,
    pub has_optional: bool,
}

#[derive(Clone, Default)]
struct PlanOpts {
    with_optional: bool,
    string_hint: Option<String>,
    realistic: bool,
}

const MAX_DEPTH: usize = 6;

/// An array field as a slice (empty when absent or not an array).
pub(crate) fn arr<'a>(v: &'a Value, key: &str) -> &'a [Value] {
    v.get(key)
        .and_then(Value::as_array)
        .map(Vec::as_slice)
        .unwrap_or(&[])
}

fn is_required(f: &Value) -> bool {
    f.get("required").and_then(Value::as_bool).unwrap_or(false)
}

fn str_field(v: &Value, key: &str) -> String {
    v.get(key).and_then(Value::as_str).unwrap_or("").to_string()
}

/// Drop the per-level stringHint before recursing (keep withOptional + realistic).
fn drop_hint(opts: &PlanOpts) -> PlanOpts {
    PlanOpts {
        with_optional: opts.with_optional,
        string_hint: None,
        realistic: opts.realistic,
    }
}

/// Resolve a TypeRef to an example value, expanding named types via the symbol
/// table. Cycle-guarded and depth-capped so recursive schemas stay finite.
fn plan_example(
    ref_: Option<&Value>,
    types: &HashMap<String, Value>,
    opts: &PlanOpts,
    seen: &HashSet<String>,
    depth: usize,
) -> ExampleValue {
    let Some(r) = ref_ else {
        return ExampleValue::Any;
    };
    if depth > MAX_DEPTH {
        return ExampleValue::Any;
    }
    if let Some(c) = r.get("const") {
        return ExampleValue::Const(c.clone());
    }
    match r.get("kind").and_then(Value::as_str) {
        Some("scalar") => scalar_example(r, opts.string_hint.as_deref(), opts.realistic),
        Some("array") => ExampleValue::Array(Box::new(plan_example(
            r.get("items"),
            types,
            &drop_hint(opts),
            seen,
            depth + 1,
        ))),
        Some("map") => ExampleValue::Map(Box::new(plan_example(
            r.get("values"),
            types,
            &drop_hint(opts),
            seen,
            depth + 1,
        ))),
        Some("any") => ExampleValue::Any,
        Some("ref") => ref_example(r, types, opts, seen, depth),
        _ => ExampleValue::Any,
    }
}

fn scalar_example(r: &Value, hint: Option<&str>, realistic: bool) -> ExampleValue {
    let fmt = r
        .get("format")
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_lowercase();
    if fmt == "binary" {
        return ExampleValue::Binary;
    }
    match r.get("scalar").and_then(Value::as_str) {
        Some("integer") => ExampleValue::Integer(if realistic { 1 } else { 0 }),
        Some("number") => ExampleValue::Number(if realistic { 1.0 } else { 0.0 }),
        Some("boolean") => ExampleValue::Boolean(true),
        _ => ExampleValue::Str(if realistic {
            realistic_string(&fmt, hint)
        } else {
            hint.unwrap_or("x").to_string()
        }),
    }
}

/// A believable string sample for a known `format`, else the hint or "string".
fn realistic_string(fmt: &str, hint: Option<&str>) -> String {
    match fmt {
        "date-time" => "2024-01-01T00:00:00Z",
        "date" => "2024-01-01",
        "email" => "user@example.com",
        "uri" | "url" => "https://example.com",
        "uuid" => "123e4567-e89b-12d3-a456-426614174000",
        "hostname" => "example.com",
        "ipv4" => "192.0.2.1",
        _ => return hint.unwrap_or("string").to_string(),
    }
    .to_string()
}

fn ref_example(
    r: &Value,
    types: &HashMap<String, Value>,
    opts: &PlanOpts,
    seen: &HashSet<String>,
    depth: usize,
) -> ExampleValue {
    let name = match r
        .get("name")
        .and_then(Value::as_str)
        .filter(|n| !n.is_empty())
    {
        Some(n) => n,
        None => return ExampleValue::Any,
    };
    let Some(named) = types.get(name) else {
        return ExampleValue::Any;
    };
    if seen.contains(name) {
        return ExampleValue::Object(Vec::new());
    }
    match named.get("kind").and_then(Value::as_str) {
        Some("enum") => {
            let val = arr(named, "values")
                .first()
                .and_then(|f| f.get("value"))
                .cloned()
                .unwrap_or_else(|| Value::String(String::new()));
            ExampleValue::Enum(val)
        }
        Some("alias") => plan_example(named.get("of"), types, &drop_hint(opts), seen, depth + 1),
        Some("union") => {
            let variant = arr(named, "variants").first();
            let mut nested = seen.clone();
            nested.insert(name.to_string());
            ExampleValue::Union(Box::new(plan_example(
                variant,
                types,
                &drop_hint(opts),
                &nested,
                depth + 1,
            )))
        }
        _ => {
            let mut nested = seen.clone();
            nested.insert(name.to_string());
            let fields = example_fields(arr(named, "fields"), types, opts, &nested, depth + 1);
            ExampleValue::Object(fields)
        }
    }
}

/// Example fields for a struct's field list; required-first, optionals only when
/// `withOptional`.
fn example_fields(
    fields: &[Value],
    types: &HashMap<String, Value>,
    opts: &PlanOpts,
    seen: &HashSet<String>,
    depth: usize,
) -> Vec<ExampleField> {
    let mut wanted: Vec<&Value> = fields
        .iter()
        .filter(|f| is_required(f) || opts.with_optional)
        .collect();
    // required first, then optional — stable so declaration order is preserved.
    wanted.sort_by_key(|f| usize::from(!is_required(f)));

    let nested = PlanOpts {
        with_optional: opts.with_optional,
        string_hint: None,
        realistic: opts.realistic,
    };
    wanted
        .into_iter()
        .map(|f| {
            let value = opts
                .realistic
                .then(|| realistic_literal(f.get("type"), &[f.get("default")]))
                .flatten()
                .unwrap_or_else(|| plan_example(f.get("type"), types, &nested, seen, depth));
            ExampleField {
                name: str_field(f, "name"),
                required: is_required(f),
                value,
            }
        })
        .collect()
}

/// The spec's own example/default for a scalar/array-typed param or field (skips
/// ref types); the neutral planner walks refs. Only reachable under `realistic`.
fn realistic_literal(type_: Option<&Value>, candidates: &[Option<&Value>]) -> Option<ExampleValue> {
    let t = type_?;
    let kind = t.get("kind").and_then(Value::as_str);
    if kind != Some("scalar") && kind != Some("array") {
        return None;
    }
    candidates
        .iter()
        .flatten()
        .find_map(|c| literal_to_example(c))
}

fn literal_to_example(v: &Value) -> Option<ExampleValue> {
    match v {
        Value::String(s) => Some(ExampleValue::Str(s.clone())),
        Value::Bool(b) => Some(ExampleValue::Boolean(*b)),
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                Some(ExampleValue::Integer(i))
            } else {
                n.as_f64().map(|f| {
                    if f.fract() == 0.0 {
                        ExampleValue::Integer(f as i64)
                    } else {
                        ExampleValue::Number(f)
                    }
                })
            }
        }
        Value::Null => Some(ExampleValue::Null),
        Value::Array(a) => a
            .first()
            .and_then(literal_to_example)
            .map(|it| ExampleValue::Array(Box::new(it))),
        _ => None,
    }
}

fn push_param(
    p: &Value,
    types: &HashMap<String, Value>,
    with_optional: bool,
    realistic: bool,
    fields: &mut Vec<ExampleField>,
) {
    let req = is_required(p);
    if !req && !with_optional {
        return;
    }
    let opts = PlanOpts {
        with_optional,
        string_hint: Some(str_field(p, "name")),
        realistic,
    };
    let value = realistic
        .then(|| realistic_literal(p.get("type"), &[p.get("example"), p.get("default")]))
        .flatten()
        .unwrap_or_else(|| plan_example(p.get("type"), types, &opts, &HashSet::new(), 0));
    fields.push(ExampleField {
        name: str_field(p, "name"),
        required: req,
        value,
    });
}

fn body_ref(method: &Value) -> Option<&Value> {
    method.get("requestBody").and_then(|b| b.get("type"))
}

fn body_has_optional(body: Option<&Value>, types: &HashMap<String, Value>) -> bool {
    let Some(br) = body else { return false };
    if br.get("kind").and_then(Value::as_str) != Some("ref") {
        return false;
    }
    let Some(name) = br.get("name").and_then(Value::as_str) else {
        return false;
    };
    types
        .get(name)
        .map(|named| arr(named, "fields").iter().any(|f| !is_required(f)))
        .unwrap_or(false)
}

/// The example call for a method: positional path args + params-struct / keyword
/// fields (query ∪ header ∪ request-body fields), required-first.
pub fn plan_method_example(
    method: &Value,
    types: &HashMap<String, Value>,
    with_optional: bool,
    realistic: bool,
) -> MethodExample {
    let path_args = arr(method, "pathParams")
        .iter()
        .map(|p| {
            let opts = PlanOpts {
                with_optional: false,
                string_hint: Some(str_field(p, "name")),
                realistic,
            };
            let value = realistic
                .then(|| realistic_literal(p.get("type"), &[p.get("example"), p.get("default")]))
                .flatten()
                .unwrap_or_else(|| plan_example(p.get("type"), types, &opts, &HashSet::new(), 0));
            PathArg {
                name: str_field(p, "name"),
                value,
            }
        })
        .collect();

    let mut fields: Vec<ExampleField> = Vec::new();
    for p in arr(method, "queryParams") {
        push_param(p, types, with_optional, realistic, &mut fields);
    }
    for p in arr(method, "headerParams") {
        push_param(p, types, with_optional, realistic, &mut fields);
    }

    // request body: flatten its struct fields into the params struct / kwargs.
    let body = body_ref(method);
    if let Some(br) = body {
        if br.get("kind").and_then(Value::as_str) == Some("ref") {
            if let Some(name) = br.get("name").and_then(Value::as_str) {
                if let Some(named) = types.get(name) {
                    if named.get("kind").and_then(Value::as_str) == Some("struct") {
                        let opts = PlanOpts {
                            with_optional,
                            string_hint: None,
                            realistic,
                        };
                        for f in
                            example_fields(arr(named, "fields"), types, &opts, &HashSet::new(), 0)
                        {
                            fields.push(f);
                        }
                    }
                }
            }
        }
    }

    // required-first for stable, readable output.
    fields.sort_by_key(|f| usize::from(!f.required));

    let has_optional = arr(method, "queryParams").iter().any(|p| !is_required(p))
        || arr(method, "headerParams").iter().any(|p| !is_required(p))
        || body_has_optional(body, types);

    MethodExample {
        path_args,
        fields,
        has_optional,
    }
}
